services/continuousquery/service.go

```diff
@@ -0,0 +1,285 @@
+		s.ReportInterval = DefaultReportTime
+	} else {
+		s.ReportInterval = interval
+	}
+
+	// Get the group by offset.
+	offset, err := cq.q.GroupByOffset()
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the CQ should be run.
+	ok, nextRun, err := cq.shouldRunContinuousQuery(now, interval)
+	if err != nil {
+		return false, err
+	} else if !ok {
+		return false, nil
+	}
+
+	// update cq.LastRun and s.lastRuns
+	cq.LastRun = nextRun
+	s.lastRuns[cqi.Name] = cq.LastRun
+
+	// Calculate and set the time range for the query.
+	// startTime should be earlier than current time.
+	startTime := nextRun.Add(-interval - offset - 1).Truncate(interval).Add(offset)
+	endTime := startTime.Add(interval - offset).Truncate(interval).Add(offset)
+
+	if err := cq.q.SetTimeRange(startTime, endTime); err != nil {
+		return false, fmt.Errorf("unable to set time range: %s", err)
+	}
+
+	// execute the query and write the results
+	res := s.runContinuousQueryAndWriteResult(cq)
+	if res.Err != nil {
+		return false, res.Err
+	}
+
+	return true, nil
+}
+
+// NewContinuousQuery returns a ContinuousQuery object with a parsed SQL statement.
+func NewContinuousQuery(database string, cqi *meta.ContinuousQueryInfo) (*ContinuousQuery, error) {
+	// Parse the query.
+	p := influxql.NewParser(strings.NewReader(cqi.Query))
+	defer p.Release()
+
+	YyParser := influxql.NewYyParser(p.GetScanner(), p.GetPara())
+	YyParser.ParseTokens()
+
+	qr, err := YyParser.GetQuery()
+	if err != nil {
+		return nil, err
+	}
+	if len(qr.Statements) == 0 {
+		return nil, nil
+	}
+	stmt := qr.Statements[0]
+
+	// Check if the statement is a valid continuous query.
+	q, ok := stmt.(*influxql.CreateContinuousQueryStatement)
+	if !ok || q.Source.Target == nil || q.Source.Target.Measurement == nil {
+		// q.Source.Target (destination) for the result of a SELECT INTO query.
+		return nil, errors.New("invalid continuous query")
+	}
+
+	// Create a new ContinuousQuery object.
+	cq := &ContinuousQuery{
+		Database: database,
+		Info:     cqi,
+		q:        q.Source,
+	}
+
+	return cq, nil
+}
+
+// shouldRunContinuousQuery returns true if the CQ should run.
+func (cq *ContinuousQuery) shouldRunContinuousQuery(now time.Time, interval time.Duration) (bool, time.Time, error) {
+	// check whether the query is an aggregate query.
+	if cq.q.IsRawQuery { // cq.q.IsRawQuery is true when is not an aggregate query.
+		return false, cq.LastRun, errors.New("continuous queries must be aggregate queries")
+	}
+
+	// Determine if we should run the continuous query based on the last time it ran.
+	if cq.HasRun { // The query has run before.
+		// Return the nextRun time for cq.
+		nextRun := cq.LastRun.Add(interval).Truncate(interval)
+		if nextRun.UnixNano() <= now.UnixNano() {
+			return true, nextRun, nil
+		}
+		return false, cq.LastRun, nil
+	} else { // The query never ran before.
+		// Retrieve the location from the CQ.
+		loc := cq.q.Location
+		if loc == nil {
+			loc = time.UTC
```

Comment:
Maybe you can add an integrated test case.

Reply:
Test case is added in `TestService_ExecuteContinuousQuery_WithTimeZone`.